mvebu: sdcard, base-files fixes and improvements

tmn505 opened the pull request:
This PR introduces small fixes, modifications and improvements to the creation of SD card image process:
fix for the sdcard partition generation,
adds bootloader configuration,
unifies creation of boot.scr,
changes bootfs to ext4,
more generic name of sysupgrade script for devices booting from sdcard.
It has also a commit retiring the usage of net-tools from base-files.

The motivation for those changes is to ease future device additions, as more and more of them rely on booting from SD cards or other "traditional" medium like hard disk. I also wanted to reduce the amount of tools necessary for this target (dosfstools, fat kmod's, net-tools).

Artox committed:
mvebu: clearfog: use partition uuid for root= bootarg

U-Boot already knows where it found the boot.scr, and
figuring out the partition UUID becomes trivial at this point.
This change allows booting OpenWrt from whatever storage it has been
flashed to: SD card, eMMC, USB disk or SATA disk.

Signed-off-by: Josua Mayer <josua.mayer97@gmail.com>
[replace lede with openwrt, redact commit message]
Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

hauke commented on target/linux/mvebu/image/Makefile:
```diff
@@ -13,8 +13,10 @@ include $(INCLUDE_DIR)/image.mk
 
 KERNEL_LOADADDR := 0x00008000
 
+SIGNATURE:=$(shell perl -e 'printf("%08x", rand(0xFFFFFFFF))')
```
This will break reproducible builds, can you generate it from SOURCE_DATE_EPOCH for example

tmn505 replied:
Good point, I'll change that. This was taken from x86.

tmn505 replied:
This is now generated from SOURCE_DATE_EPOCH as recommended.

hauke commented on target/linux/mvebu/image/Makefile:
```diff
@@ -195,7 +196,7 @@ define Device/armada-388-clearfog-pro
   KERNEL := kernel-bin
   DEVICE_TITLE := SolidRun ClearFog Pro
-  DEVICE_PACKAGES := mkf2fs e2fsprogs swconfig kmod-fs-vfat kmod-nls-cp437 kmod-nls-iso8859-1
+  DEVICE_PACKAGES := mkf2fs e2fsprogs swconfig
```
is ext4 support always included by default?

tmn505 replied:
Ext4 is a built-in in this target https://github.com/openwrt/openwrt/blob/adaf1cbcc8b253ea807dbe0416b4b04c33dceadf/target/linux/mvebu/config-4.9#L159.

KanjiMonster replied:
Have you verified this works? IIRC the clearfog u-boot loads the kernel from the vfat partition and I don't see any u-boot changes.

Also does this affect sysupgrade with keeping config? The vfat partition is used for temporary storage of the saved config, can this cause issues when upgrading from a 17.01 system?

tmn505 replied:
I don't have clearfog, so can't test it if it works there. The u-boot for clearfog has ext4 enabled and is using distro boot compatibility (set of autodetecting commands, see the bootscripts, there is no mention of fat), so should work without problems.

The sysupgrade was tested with espressobin with the same sd card image layout (without bootloader), and it went smoothly.

I'll try to ping the author of the changes which my commits touch and ask him if he can test them.

KanjiMonster replied:
I gave it a test, and while it boots fine with the new system,  a sysupgrade from LEDE 17.01 will fail to backup the configuration:
```
root@testname:/# sysupgrade -v /tmp/openwrt-mvebu-armada-388-clearfog-pro-sdcard.img.gz
Saving config files...
etc/config/dhcp
etc/config/dropbear
etc/config/firewall
etc/config/luci
etc/config/network
etc/config/rpcd
etc/config/system
etc/config/ubootenv
etc/config/ucitrack
etc/config/uhttpd
etc/dropbear/dropbear_rsa_host_key
etc/group
etc/hosts
etc/inittab
etc/luci-uploads/.placeholder
etc/opkg/keys/5151f69420c3f508
etc/opkg/keys/72a57f2191b211e0
etc/opkg/keys/792d9d9b39f180dc
etc/opkg/keys/9ef4694208102c43
etc/opkg/keys/b5043e70f9a75cde
etc/opkg/keys/dace9d4df16896bf
etc/opkg/keys/dd6de0d06bbd3d85
etc/passwd
etc/profile
etc/rc.local
etc/shadow
etc/shells
etc/sysctl.conf
etc/sysctl.d/local.conf
killall: watchdog: no process killed
Sending TERM to remaining processes ... uhttpd ntpd dnsmasq ubusd logd rpcd netifd odhcpd
Sending KILL to remaining processes ...
Switching to ramdisk...
Performing system upgrade...
0+155 records in
0+155 records out
mount: mounting /dev/mmcblk0p1 on /boot failed: Invalid argument
umount: can't unmount /boot: Invalid argument
Upgrade completed
Rebooting system...
(...)
root@OpenWrt:/#
```

As you can see the old system fails to mount the new /boot as it tries to mount it as vfat. This breaks keeping the configuration, which is bad IMHO.

tmn505 replied:
This didn't occur to me previously, You'r right. I'll keep the packages and modify the sysupgrade script. After the next stable is out, I'll send PR to remove fat leftovers.

tmn505 commented:
@Artox Sorry for bothering. Most commits are touching Your changes, can You comment or/and test them on bare metal?

Artox commented:
@tmn505 
My changes are minimal and only affect partition uuid. Perhaps you can just rebase on top my commit doing that?

tmn505 committed:
mvebu: fix partition type and signature for sdcard

Previously the partition signature was assigned from provided type. Now
both are corrected wherein signature is always generated from
SOURCE_DATE_EPOCH. With that the root file system can be identified
by PARTUUID string, without relying on static declaration of device node.
This commit also does some cosmetics, removing trailing whitespace and
replacing spaces with tab.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

tmn505 committed:
mvebu: rename clearfog sysupgrade script and functions

Cosmetic name change of ClearFog sysupgrade script and its functions to
more generic sdcard. This way it won't be confusing for other future
device additions.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

tmn505 committed:
mvebu: base-files: moving on from net-tools

Remove dependency on net-tools, replacing calls to ifconfig with iproute
equivalent, in functions changing interface MAC addresses.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

tmn505 commented:
Additional commits:
from Artox to identify rootfs by uuid, it's better approach when U-Boot uses distro boot compatibility;
(fix) remove the redefinition of IMAGE_NAME.
Changes:
removed rootfs identification by SIGNATURE, for now nothing will be using it so no reason to add dead code;
boot.scr is now optional on image creation;
added unmount of boot when moving sysupgrade saved configs.

@KanjiMonster please test, copying saved configs on sysupgrade should be fixed.

@Artox I left out the U-Boot upgrade, I don't have ClearFog so can't test it.

Artox commented:
@mnr Very good!
Concerning the u-boot version: My colleague said 2018.01 works very well, in particular with booting from USB!
So perhaps @KanjiMonster wants to take a look.

KanjiMonster commented:
> @KanjiMonster please test, copying saved configs on sysupgrade should be fixed.

Still doesn't work. The main issue is that 17.01's `sysupgrade` tries to mount the boot partition as vfat, and nothing else, and there is nothing you can do to change that. So if we do want to provide a smooth upgrade, we need to defer the switch to ext4 to after the next release.

tmn505 commented:
@KanjiMonster You're completely right, i mindlessly removed `-t vfat` before creating fat image in my simulation, that made it run without problems.
I'll defer the transition. Thanks.

@commodo Great that You started to work on this. I'll test this today or tomorrow, and drop few pointers after that.

tmn505 committed:
mvebu: remove fs declaration in mount commands

Allows to have other file system for boot partition without breaking
sysupgrade.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

tmn505 commented:
@hauke @KanjiMonster 
The changes are complete, please review/comment them.

Hauke, for now it's not possible to switch to only ext4 for ClearFog devices using SD card without breaking sysupgrade, I'll do the switch after the next stable is out. For now when introducing ESPRESSObin support I'll add another define to create ext4 only SD car image for it.

hauke commented on target/linux/mvebu/image/Makefile:
```diff
@@ -25,22 +25,17 @@ endef
 	$(foreach dts,$(DEVICE_DTS), mcopy -i $@.boot $(DTS_DIR)/$(dts).dtb ::$(dts).dtb;)
 	mcopy -i $@.boot $(IMAGE_KERNEL) ::zImage
-	mcopy -i $@.boot $@.bootscript ::boot.scr
+	-mcopy -i $@.boot $@-boot.scr ::boot.scr
```
Why is this allowed to fail now?

tmn505 replied:
Seems like not every device uses `boot.scr` file, look at PR #684. That way the `boot-scr-dummy` define in mentioned PR won't be necessary.

hauke reviewed (approved):
This looks good to me, Can you please rebase this on top of current master.
I would like to take this soon, probably replacing the sdcard code with a common version.

tmn505 committed:
mvebu: remove redefinition of image name for clearfog

The IMAGE_NAME redefinition causes overwriting of generated SD card
image when multiple root file system types are selected. In result only
single SD card image is generated. This commit fixes this behaviour.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

tmn505 committed:
mvebu: make sdcard bootloader option configurable

Remove the necessity for boot loader from SD card image creation process
and make it configurable.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

tmn505 committed:
mvebu: unify boot.scr creation

Unify boot.scr generation so Makefile for device image generation won't
grow without a reason. Also make boot-scr step optional.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>

hauke commented:
@KanjiMonster Does this look ok for you now?

hauke commented:
Thank you for the patches, they are now applied to master.
